### ISSUE TITLE
Is a graph of operations the right level of abstraction for a web standard?

### ISSUE THREAD

**Issue by jbingham** [w3c/machine-learning-charter#2]
Given the rapid evolution of operations, and how fast ML as a field is evolving, is it premature to create a working group to propose a standard that’s expected to endure forever?

The number of operations defined in TensorFlow, ONNX, PyTorch and other machine learning frameworks has been growing at double-digit rates every year. Meanwhile, incompatible changes have led to new versions of some operations (eg, suffixed with v2) while others have fallen into disuse.

At Google, the Android and TensorFlow teams have experience shipping a neural network API to a billion devices. They are moving away from an operation-based API along the lines of what the Web NN API is proposing. For a little more context, see this [FAQ](https://github.com/webmachinelearning/model-loader/blob/master/explainer.md#why-not-use-a-graph-api-like-the-neural-network-api-nn-api-on-android) in the Model Loader API Explainer. 

Should this work continue to incubate in the community group until the field of ML stabilizes a bit more, or at least there’s wider agreement about a stable path forward?

Even without new APIs, ML on the web has been rapidly improving, as JavaScript libraries have added support for WebGL and WASM. Developers are only beginning to take advantage of the capabilities that exist today. We’d love for them to get access to the latest hardware acceleration too, in a way that makes sense for the long-term.

**Comment by anssiko** [comment]
It seems issue #7 has received significant attention and has already started discussing the following points that were raised in this issue (recasted in my own words, possible interpretation errors mine):
- Graph of operations abstraction return on investment
- Feasibility of maintaining such an abstraction when the number of operations is growing rapidly
- How to accommodate the still evolving lower-level instruction set proposals

@jbingham, considering ongoing discussion in #7, do you think there is an additional perspective that should be discussed specifically in this issue, or could we close this issue and continue discussion in #7 on these points?

I understand the title of this issue ("Is a graph of operations the right level of abstraction for a web standard?") asks an explicit question. I believe the proposed PR #9 to add Web Neural Network API Lower-level Instruction Set as a tentative specification would give the WG the needed flexibility to allow that work to advance to the WG depending on the progress.

**Comment by jbingham** [comment]
Yes, PR #9 provides a tentative spec that accommodates a lower level set of instructions. The question here is about the right mix of higher level ops and lower level instructions. Note that this is independent of whether the API is a graph construction or model loader API.

#7 does have some relevant information, like the observation that the majority of higher level TensorFlow ops have fallen into disuse, which would be very undesirable for a web standard. Also, of the in-use operations, many of them can be decomposed into lower level instructions. 

Concretely, it sounds like we're in agreement about some guiding principles: eg, let's decompose into lower level instructions where possible, let's try to bound the number of operations so it doesn't grow too large. Maybe that's the answer, and if others agree, we can close this issue.

**Comment by anssiko** [comment]
Per discussion on the [WebML CG Teleconference – 10 December 2020](https://www.w3.org/2020/12/10-webmachinelearning-minutes.html) this issue can be closed.
